plutus-use-cases/src/Plutus/Contracts/Governance.hs

```diff
@@ -0,0 +1,199 @@
+
+instance AsSMContractError GovError where
+    _SMContractError = _GovStateMachineError
+
+type GovernanceMachine = StateMachine GovState GovInput
+
+{-# INLINABLE machine #-}
+machine :: Params -> GovernanceMachine
+machine params = SM.mkStateMachine (transition params) isFinal where
+    {-# INLINABLE isFinal #-}
+    isFinal _ = False
+
+{-# INLINABLE mkValidator #-}
+mkValidator :: Params -> Scripts.ValidatorType GovernanceMachine
+mkValidator params = SM.mkValidator $ machine params
+
+scriptInstance :: Params -> Scripts.ScriptInstance GovernanceMachine
+scriptInstance = Scripts.validatorParam @GovernanceMachine
+    $$(PlutusTx.compile [|| mkValidator ||])
+    $$(PlutusTx.compile [|| wrap ||])
+    where
+        wrap = Scripts.wrapValidator
+
+client :: Params -> SM.StateMachineClient GovState GovInput
+client params = SM.mkStateMachineClient $ SM.StateMachineInstance (machine params) (scriptInstance params)
+
+{-# INLINABLE votingValue #-}
+votingValue :: MonetaryPolicyHash -> TokenName -> Value.Value
+votingValue mph tokenName =
+    Value.singleton (Value.mpsSymbol mph) tokenName  1
+
+{-# INLINABLE ownsVotingToken #-}
+ownsVotingToken :: MonetaryPolicyHash -> TokenName -> TxConstraints Void Void
+ownsVotingToken mph tokenName = mempty -- TODO
+
+{-# INLINABLE transition #-}
+transition :: Params -> State GovState -> GovInput -> Maybe (TxConstraints Void Void, State GovState)
+transition Params{..} State{ stateData = s, stateValue = currentValue} i = case (s, i) of
+
+    (GovState{mph}, ForgeTokens) ->
+        let (total, constraints) = foldMap (\pk -> let v = votingValue mph tokenName in (v, Constraints.mustPayToPubKey pk v)) initialHolders
+        in Just (constraints <> Constraints.mustForgeValue total, State s currentValue)
+
+    (GovState law mph Nothing, ProposeChange proposal) ->
+        Just (ownsVotingToken mph tokenName, State (GovState law mph (Just (Voting proposal AssocMap.empty))) currentValue)
+
+    (GovState law mph (Just (Voting p oldMap)), AddVote pk vote) ->
+        let newMap = AssocMap.insert pk vote oldMap
+            constraints = ownsVotingToken mph tokenName <> Constraints.mustBeSignedBy pk
+        in Just (constraints, State (GovState law mph (Just (Voting p newMap))) currentValue)
+
+    (GovState oldLaw mph (Just (Voting Proposal{newLaw} votes)), FinishVoting) ->
+        let (Sum ayes, Sum nays) = foldMap (\b -> if b then (Sum 1, Sum 0) else (Sum 0, Sum 1)) votes
+        in if ayes >= requiredVotes -- Enough votes in favor
+            then Just (mempty, State (GovState newLaw mph Nothing) currentValue)
+            else if nays > length initialHolders - requiredVotes -- Enough opposed votes
+                then Just (mempty, State (GovState oldLaw mph Nothing) currentValue)
+                else Nothing -- Not enough votes either way, use cancel-voting to cancel
+
+    (GovState law mph (Just _), CancelVoting) ->
+        Just (mempty, State (GovState law mph Nothing) currentValue)
```

Comment:
if you want to use the deadline here, you can put it into a constraint, replacing `empty` with

```haskell
Constraints.mustValidateIn (Interval.from deadline)
```